src/osd/OSD.cc

```diff
@@ -5899,52 +5896,48 @@ void OSD::sched_scrub()
 {
   assert(osd_lock.is_locked());
 
-  bool load_is_low = scrub_should_schedule();
-
-  dout(20) << "sched_scrub load_is_low=" << (int)load_is_low << dendl;
-
   utime_t now = ceph_clock_now(cct);
+
+  bool load_is_low = scrub_load_below_threshold();
+  bool time_permit = scrub_time_permit(now);
+  dout(20) << __func__ << " load_is_low=" << (int)load_is_low << dendl;
   
   //dout(20) << " " << last_scrub_pg << dendl;
 
   pair<utime_t, spg_t> pos;
   if (service.first_scrub_stamp(&pos)) {
     do {
-      utime_t t = pos.first;
-      spg_t pgid = pos.second;
-      dout(30) << "sched_scrub examine " << pgid << " at " << t << dendl;
-
-      utime_t diff = now - t;
-      if ((double)diff < cct->_conf->osd_scrub_min_interval) {
-	dout(10) << "sched_scrub " << pgid << " at " << t
-		 << ": " << (double)diff << " < min (" << cct->_conf->osd_scrub_min_interval << " seconds)" << dendl;
-	break;
-      }
-      if ((double)diff < cct->_conf->osd_scrub_max_interval && !load_is_low) {
+      const utime_t& sched_time = pos.first;
+      const spg_t& pgid = pos.second;
+      dout(30) << "sched_scrub examine " << pgid << " at " << sched_time << dendl;
+
+      double delay = now - sched_time;
+      if (delay < 0) {
 	// save ourselves some effort
-	dout(10) << "sched_scrub " << pgid << " high load at " << t
-		 << ": " << (double)diff << " < max (" << cct->_conf->osd_scrub_max_interval << " seconds)" << dendl;
+	dout(10) << "sched_scrub " << pgid << " schedued at " << sched_time
+		 << " > " << now << dendl;
 	break;
       }
 
       PG *pg = _lookup_lock_pg(pgid);
-      if (pg) {
-	if (pg->get_pgbackend()->scrub_supported() && pg->is_active() &&
+      if (!pg)
+	continue;
+      if (pg->get_pgbackend()->scrub_supported() && pg->is_active() &&
+	  (pg->scrubber.must_scrub ||
+	   (time_permit &&
 	    (load_is_low ||
-	     (double)diff >= cct->_conf->osd_scrub_max_interval ||
-	     pg->scrubber.must_scrub)) {
-	  dout(10) << "sched_scrub scrubbing " << pgid << " at " << t
-		   << (pg->scrubber.must_scrub ? ", explicitly requested" :
-		   ( (double)diff >= cct->_conf->osd_scrub_max_interval ? ", diff >= max" : ""))
-		   << dendl;
-	  if (pg->sched_scrub()) {
-	    pg->unlock();
-	    break;
-	  }
+	     delay >= cct->_conf->osd_scrub_max_interval)))) {
```

Comment:
this is no longer quite right, probably delay >= (osd_scrub_max_interval - osd_scrub_min_interval) or include the max_scrub scheduling time explicitely in the queue?